File: server/src/controllers/machineController.js
import prisma from "../prismaClient.js";

export const getMachines = async (req, res) => {
    try {
        const userEmail = req.user.email;
        let machines = [];

        try {
            if (userEmail === "admin@rafcloud") {
                try {
                    machines = await prisma.machine.findMany({
                        include: {
                            createdBy: { select: { id: true, fullName: true, email: true } },
                            // errors: true, 
                        },
                    });
                } catch (err) {
                    console.error("Prisma error (admin fetch):", err);
                    machines = [];
                }
            } else {
                machines = await prisma.machine.findMany({
                    where: { createdById: req.user.id },
                    include: {
                        createdBy: { select: { id: true, fullName: true, email: true } },
                        // errors: true,
                    },
                });
            }
        } catch (prismaErr) {
            console.error("Prisma error:", prismaErr);
            machines = []; // fallback
        }

        res.json(machines);
    } catch (err) {
        console.error("Unexpected error:", err);
        res.status(500).json({ error: "Failed to fetch machines" });
    }
};

export const createMachine = async (req, res) => {
    try {
        const { name } = req.body;
        const userId = req.user.id;

        if (!name) {
            return res.status(400).json({ error: "Name is required" });
        }

        const newMachine = await prisma.machine.create({
            data: {
                name,
                state: "Off", // default state
                createdById: userId,
            },
        });

        res.json(newMachine);
    } catch (err) {
        console.error(err);
        res.status(500).json({ error: "Failed to create machine" });
    }
};

export const getMachineById = async (req, res) => {
    try {
        const id = parseInt(req.params.id);
        const machine = await prisma.machine.findUnique({
            where: { id },
            include: { createdBy: true, errors: true },
        });

        if (!machine) return res.status(404).json({ error: "Machine not found" });
        res.json(machine);
    } catch (err) {
        console.error(err);
        res.status(500).json({ error: "Failed to fetch machine" });
    }
};

// Turn On
export const turnOn = async (req, res) => {
    try {
        const { id } = req.params;

        const machine = await prisma.machine.update({
            where: { id: parseInt(id) },
            data: { state: "On" },
        });

        const newError = await prisma.errorLog.create({
            data: {
                message: "Error while turning on",
                machineId: parseInt(id),
            },
        });

        res.json({ message: "Machine turned on", machine, error: newError });
    } catch (err) {
        console.error(err);
        res.status(500).json({ error: "Failed to turn on machine" });
    }
};

// Turn Off
export const turnOff = async (req, res) => {
    try {
        const { id } = req.params;

        const machine = await prisma.machine.update({
            where: { id: parseInt(id) },
            data: { state: "Off" },
        });

        const newError = await prisma.errorLog.create({
            data: {
                message: "Error while turning off",
                machineId: parseInt(id),
            },
        });

        res.json({ message: "Machine turned off", machine, error: newError });
    } catch (err) {
        console.error(err);
        res.status(500).json({ error: "Failed to turn off machine" });
    }
};

// Restart
export const restart = async (req, res) => {
    try {
        const { id } = req.params;
        const newError = await prisma.errorLog.create({
            data: {
                message: "Error while restarting",
                machineId: parseInt(id),
            },
        });
        res.json({ message: "Error logged for restarting", error: newError });
    } catch (err) {
        console.error(err);
        res.status(500).json({ error: "Failed to log error" });
    }
};

// Destroy
export const destroy = async (req, res) => {
    try {
        const { id } = req.params;

        const machine = await prisma.machine.update({
            where: { id: parseInt(id) },
            data: { active: false },
        });

        const newError = await prisma.errorLog.create({
            data: {
                message: "Error while destroying",
                machineId: parseInt(id),
            },
        });

        res.json({ message: "Machine destroyed", machine, error: newError });
    } catch (err) {
        console.error(err);
        res.status(500).json({ error: "Failed to destroy machine" });
    }
};

File: demo/src/main/java/com/example/demo/controllers/MachineController.java
package com.example.demo.controllers;

import com.example.demo.model.Machine;
import com.example.demo.requests.CreateMachineRequest;
import com.example.demo.security.AuthContext;
import com.example.demo.security.RequiredPermission;
import com.example.demo.services.MachineService;
import jakarta.validation.Valid;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.Map;

@RestController
@RequestMapping("/api/machines")
public class MachineController {

    private final MachineService machineService;

    public MachineController(MachineService machineService) {
        this.machineService = machineService;
    }

    @RequiredPermission("create_machine")
    @PostMapping
    public ResponseEntity<?> create(@Valid @RequestBody CreateMachineRequest req) {
        var auth = AuthContext.get();
        if (auth == null) return ResponseEntity.status(401).body(Map.of("error", "No token provided"));

        Machine m = machineService.createMachine(req.name, auth.id);
        if (m == null) return ResponseEntity.status(500).body(Map.of("error", "Failed to create machine"));
        return ResponseEntity.ok(m);
    }

    @RequiredPermission("search_machine")
    @GetMapping
    public ResponseEntity<?> list() {
        var auth = AuthContext.get();
        if (auth == null) return ResponseEntity.status(401).body(Map.of("error", "No token provided"));

        return ResponseEntity.ok(machineService.getMachines(auth.id, auth.email));
    }

    @RequiredPermission("search_machine")
    @GetMapping("/{id}")
    public ResponseEntity<?> byId(@PathVariable Long id) {
        var auth = AuthContext.get();
        if (auth == null) return ResponseEntity.status(401).body(Map.of("error", "No token provided"));

        Machine m = machineService.getByIdForUser(id, auth.id, auth.email);
        if (m == null) return ResponseEntity.status(404).body(Map.of("error", "Machine not found"));
        return ResponseEntity.ok(m);
    }

    @RequiredPermission("turnon_machine")
    @PostMapping("/{id}/turn-on")
    public ResponseEntity<?> turnOn(@PathVariable Long id) {
        var auth = AuthContext.get();
        if (auth == null) return ResponseEntity.status(401).body(Map.of("error", "No token provided"));

        Machine m = machineService.turnOnForUser(id, auth.id, auth.email);
        if (m == null) return ResponseEntity.status(404).body(Map.of("error", "Machine not found"));
        return ResponseEntity.ok(Map.of("message", "Turning on operation started", "machine", m));
    }

    @RequiredPermission("turnoff_machine")
    @PostMapping("/{id}/turn-off")
    public ResponseEntity<?> turnOff(@PathVariable Long id) {
        var auth = AuthContext.get();
        if (auth == null) return ResponseEntity.status(401).body(Map.of("error", "No token provided"));

        Machine m = machineService.turnOffForUser(id, auth.id, auth.email);
        if (m == null) return ResponseEntity.status(404).body(Map.of("error", "Machine not found"));
        return ResponseEntity.ok(Map.of("message", "Turning off operation started", "machine", m));
    }

    @RequiredPermission("restart_machine")
    @PostMapping("/{id}/restart")
    public ResponseEntity<?> restart(@PathVariable Long id) {
        var auth = AuthContext.get();
        if (auth == null) return ResponseEntity.status(401).body(Map.of("error", "No token provided"));

        boolean ok = machineService.restartForUser(id, auth.id, auth.email);
        if (!ok) return ResponseEntity.status(404).body(Map.of("error", "Machine not found"));
        return ResponseEntity.ok(Map.of("message", "Error logged for restarting"));
    }

    @RequiredPermission("destroy_machine")
    @PostMapping("/{id}/destroy")
    public ResponseEntity<?> destroy(@PathVariable Long id) {
        var auth = AuthContext.get();
        if (auth == null) return ResponseEntity.status(401).body(Map.of("error", "No token provided"));

        Machine m = machineService.destroyForUser(id, auth.id, auth.email);
        if (m == null) return ResponseEntity.status(404).body(Map.of("error", "Machine not found"));
        return ResponseEntity.ok(Map.of("message", "Machine destroyed", "machine", m));
    }
}